README.md
swank-js
========

** IMPORTANT NOTE **
The primary swank-js source repository is now [https://github.com/swank-js/swank-js](https://github.com/swank-js/swank-js).


swank-js provides [SLIME](http://common-lisp.net/project/slime/) REPL
and other development tools for in-browser JavaScript and
[Node.JS](http://nodejs.org). It consists of SWANK backend and
accompanying SLIME contrib. [Socket.IO](http://socket.io/) is used to
communicate with wide range of web browsers.

Motivation
----------

From my experience an ability to use REPL for JavaScript debugging and
interactive development is very helpful when developing Web
applications. Previously I was using a heavily patched
[MozRepl](https://github.com/bard/mozrepl/wiki/) version that was
adapted for in-browser JavaScript. Primary downsides of that approach
were extreme instability of communication between Emacs and the
browser, the lack of cross-browser support and the lack of good RPC
between Emacs and JS that can be used to develop some handy
extensions.

I knew there exists [slime-proxy](https://github.com/3b/slime-proxy)
project that provides such functionality for
[Parenscript](http://common-lisp.net/project/parenscript/). The
problem is that most of us including myself can't use Lisp all the
time and a lot of code needs to be developed using languages like
plain JavaScript (as opposed to Parenscript), Python and so on. My
first thought was to adapt slime-proxy for use with plain JS, but then
I decided to roll my own SWANK backend using Node.JS. I wanted to find
out what this buzz around Node.JS is about and perhaps steal an
idea or two from there for use in my Lisp projects. Another reason was
availability of [Socket.IO](http://socket.io/) and an example of
[tiny http server proxy](http://www.catonmat.net/http-proxy-in-nodejs).

Some people may prefer Firebug or built-in browser development tools
to Emacs-based development, but for example in case of mobile browsers
you don't have much choice. At some point I did try swank-js with an
colleague's iPhone and it worked, which is not too unexpected given
that Socket.IO has good cross-browser support.

Status
------

As of now swank-js provides REPL with an ability to work with multiple
browser connections, supports dynamic updates of JavaScript code using
C-c C-c / C-M-x, provides debug output function and an ability to
reload web pages in the browser or refresh their CSS using Emacs
commands.

Many aspects of full-fledged SWANK backends aren't implemented yet,
there's no debugger/completion/autodoc and so on, but as I plan to use
swank-js a lot in future there's a good chance many of these features
will be eventually added.

Installation
------------

1. Install [Node.JS](http://nodejs.org) and [npm](http://npmjs.org/)

2. Install swank-js from npm:

        npm install -g swank-js

3. Get recent [SLIME](http://common-lisp.net/project/slime/) from its CVS
or the [git mirror](http://git.boinkor.net/gitweb/slime.git). The backend
was verified to work with SLIME 2012-02-12, it may or may not work with
other versions, but note that breaking change in the protocol was introduced
in SLIME 2011-11-27.

4. Make sure you have latest [js2-mode](http://code.google.com/p/js2-mode/).
Add it to your .emacs:

        (add-to-list 'load-path "/path/to/js2-mode/directory")
        (autoload 'js2-mode "js2-mode" nil t)
        (add-to-list 'auto-mode-alist '("\\.js$" . js2-mode))

5. Create symbolic link to slime-js.el in the contrib subdirectory of
SLIME project.

6. Install [js2-mode](http://code.google.com/p/js2-mode/) into emacs from
   http://tromey.com/elpa/

7. In your .emacs, add the following lines (you may use other key for
slime-js-reload; also, if you're already using SLIME, just add slime-js
to the list of contribs, otherwise adjust the load-path item):

        (global-set-key [f5] 'slime-js-reload)
        (add-hook 'js2-mode-hook
                  (lambda ()
                    (slime-js-minor-mode 1)))

8. If you're using CSS mode, you may want to add the following lines too:

        (add-hook 'css-mode-hook
                  (lambda ()
                    (define-key css-mode-map "\M-\C-x" 'slime-js-refresh-css)
                    (define-key css-mode-map "\C-c\C-r" 'slime-js-embed-css)))

Usage
-----

If you want to use swank from the node project just add following to your
package.json file:

        "devDependencies": {
          "swank-js": ">=0.0.1"
        },
        "scripts": {
          "swank": "node node_modules/swank-js"
        }

Once this is done you should be able to run up a swank for this project by
running:

        npm run swank

Alternatively you can install swank-js globally by running:

        npm install -g swank-js

Once installed you could run it from you project directory:

        swank-js

Make SLIME connect to the backend using `M-x slime-connect` and
specifying `localhost` and port `4005`. You will see REPL buffer
with the following prompt:

    NODE>

This means that you're currently talking to Node.JS. You may play
around with it by running some JavaScript expressions.

If you get warning about SLIME version mismatch, you may make it
disappear until the next SLIME upgrade by typing *,js-slime-version*
at the REPL and entering your SLIME version (e.g. 2010-11-13).

### Connecting to a web browser ###

Point your web browser to

    http://localhost:8009/swank-js/test.html
You will see the following message appear in the REPL (browser name
and version may differ):

    Remote attached: (browser) Firefox3.6:127.0.0.1

This means that the browser is now connected. Several browsers can
connect simultaneously and you can switch between them and Node.JS
REPL using *,select-remote* REPL shortcut. To use it, press ','
(comma) and type *select-remote* (completion is supported). You will
see "Remote:" prompt. Press TAB to see completions. Select your
browser in the list by typing its name or clicking on the
completion. The following message will appear:

    NODE>
    Remote selected: (browser) Firefox3.6:127.0.0.1
    FIREFOX-3.6>

After that, you can interactively evaluate expressions in your
browser. To go back to Node.JS repl, switch back to node.js/direct
remote.

    FIREFOX-3.6> document.body.nodeName
    BODY
    FIREFOX-3.6> alert("test!")
    FIREFOX-3.6>

When working with browser, you may use F5 to reload the page. swank-js
connection with browser is lost in this case, but to solve this
problem you may use *,sticky-select-remote* instead of
*,select-remote*. This way swank-js will remember your selection and
automagically attach to the browser whenever it connects. If you press
F5 after using *,sticky-select-remote*, you will see that browser
briefly disconnects but then connects again:

    Remote detached: (browser) Firefox3.6:127.0.0.1
    FIREFOX-3.6>
    Remote selected (auto): (direct) node.js
    Remote attached: (browser) Firefox3.6:127.0.0.1
    NODE>
    Remote selected (auto): (browser) Firefox3.6:127.0.0.1
    FIREFOX-3.6>

The sticky remote selection is saved in the config file, ~/.swankjsrc,
so you don't need to do *,sticky-select-remote* after restarting the
browser.


### Connecting to a remote page ###

Now, let's try to make it work with an actual site. swank-js acts as a
proxy between your browser and the site so it can inject necessary
script tags into HTML pages and avoid cross-domain HTTP request
problems. Let's point it to [reddit](http://www.reddit.com). Type
*,target-url* and then *http://www.reddit.com* (www. part is
important, otherwise it will redirect to www.reddit.com skipping the
proxy). Point your browser to http://localhost:8009, you'll see reddit
frontpage load. If you didn't do *,select-remote* or
*,sticky-select-remote* yet do it now and select your browser from the
list of remotes. Now you can execute JavaScript in the context of
reddit:

    FIREFOX-3.6> $(".sitetable a.title").map(function(n) { return (n + 1) + ". " + $(this).text(); }).get().join("\n")
    1. Wikileaks currently under a mass DDOS attack
    2. Munich University - Jealous
    ...

Let's make a function from it. Create a file test.js somewhere and
make sure it uses js2-mode (if it doesn't, switch it to js2-mode using
M-x js2-mode). Type the following there:

    function listRedditTitles () {
      $(".sitetable a.title").map(
        function (n) {
          SwankJS.output((n + 1) + ". " + $(this).text() + "\n");
        }).get().join("\n");
    }

Note SwankJS.output() function being used there. It allows you to send
debug print to SLIME REPL.

Move the point somewhere into the middle of the listRedditTitles() function
and press C-M-x. Now you may try it out in the REPL:

    FIREFOX-3.6> listRedditTitles()
    1. Wikileaks currently under a mass DDOS attack
    2. Munich University - Jealous
    ...

You may edit the function definition and update it using C-M-x any
number of times.


### Hacking CSS ###

#### By updating a file ####

Now let's try some CSS hacking. Create a directory named zzz and start
a Web server in it from your command prompt:

    $ mkdir zzz && cd zzz && python -mSimpleHTTPServer

Create a file named a.css there and make sure it uses css-mode (like
with js2-mode, you can force it with M-x css-mode). Add some CSS rules
to this file:

    body {
        background: green;
    }

Now let's add the stylesheet to the reddit page:

    FIREFOX-3.6> $('head').append('<link rel="stylesheet" href="http://localhost:8000/a.css" type="text/css" />');
    [object Object]

You will see some parts of the page become green. Now, change green to
blue in the CSS file and press C-M-x (it will save the file
automatically):

    body {
        background: blue;
    }

You will see the page become blue, maybe after some flickering as all
CSS used on the page is reloaded. This way you may update CSS in an
AJAX application without reloading the page, which is often rather
handy. Unlike editing CSS in Firebug in case when you're editing CSS
of your own application changes will not disappear upon page reload
(with reddit page you'll have to readd the stylesheet).


#### By embedding CSS ####

Alternatively to just try out a snippet of CSS you can select some CSS code and hit `C-c C-r`. This will send the code snippet (or the content of the whole buffer) to the browser and embed it inside a style element.

To remove the embedded CSS run the command with a prefix `C-u C-c C-r`.


### Embedding swank-js in a page ###

This is useful for automatically connecting to a web page you develop
locally without using the *,target-url* command and without changing
the document URL for that page. When `node swank.js` is running embed

    <script type="text/javascript" src="http://localhost:8009/swank-js/swank-js-inject.js"></script>

and you are ready to go!

### Swank js as a bookmarklet ###

Now as easy as running a bookmarklet!  Just add a bookmark, and set this as the url:

    javascript:(function(d)%7Bwindow.swank_server%3D%27http%3A%2F%2Flocalhost%3A8009%2F%27%3Bif(!d.getElementById(%27swank-js-inj%27))%7Bvar%20h%3Dd.getElementsByTagName(%27head%27)%5B0%5D%2Cs%3Dd.createElement(%27script%27)%3Bs.id%3D%27swank-js-inj%27%3Bs.type%3D%27text%2Fjavascript%27%3Bs.src%3Dswank_server%2B%27swank-js%2Fswank-js-inject.js%27%3Bh.appendChild(s)%3B%7D%7D)(document)%3B

To disconnect you can use this bookmarklet:

    javascript:(function()%7BSwankJS.disconnect()%3B%7D)()%3B


Troubleshooting
---------------

I've noticed that flashsocket Socket.IO transport does exhibit some
instability. You may want to try other transports by changing the
socketio cookie, e.g.:

    document.cookie = "socketio=xhr-polling"

Be careful not to lose connection to the browser though especially in
case of REPL-less browser like IE6/7 or you'll have to type something
like

    javascript:void(document.cookie = "socketio=flashsocket")

in the address bar.

In case of IE, increasing the maximum number of HTTP connections may
help with non-Flash transports, although I didn't try it yet. To do it
add DWORD value MaxConnectionsPer1_0Server to the following registry
key:

    HKEY\_CURRENT\_USER\Software\Microsoft\Windows\CurrentVersion\Internet Settings

License
-------

swank-js is distributed under X11-style license. See LICENSE file.


History.md
# Changes

## 0.0.3 / 2012-03-23

    - In require resolve relative module ids to a working directory.
    - Add an option for reloading modules `require('foo', { reload: true })`.

## 0.0.2 / 2012-03-23

    - Update project metadata to point to a new repository.

## 0.0.1 / 2012-02-26

    - Initial release.


swank.js
#!/usr/bin/env node
// -*- mode: js2 -*-
//
// Copyright (c) 2010 Ivan Shvedunov. All rights reserved.
// Copyright (c) 2012 Robert Krahn. All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions
// are met:
//
// * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//
// * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following
// disclaimer in the documentation and/or other materials
// provided with the distribution.
//
// THIS SOFTWARE IS PROVIDED BY THE AUTHOR 'AS IS' AND ANY EXPRESSED
// OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
// GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

var net = require("net"), http = require('http'), io = require('socket.io'), util = require("util"),
    url = require('url'), fs = require('fs');
var swh = require("./swank-handler");
var swp = require("./swank-protocol");
var ua = require("./user-agent");
var config = require("./config");

var DEFAULT_TARGET_HOST = "localhost";
var DEFAULT_TARGET_PORT = 8080;
var CONFIG_FILE_NAME = "~/.swankjsrc";

var cfg = new config.Config(CONFIG_FILE_NAME);
var executive = new swh.Executive({ config: cfg });

var swankServer = net.createServer(
  function (stream) {
    var handler = new swh.Handler(executive);
    var parser = new swp.SwankParser(
      function onMessage (message) {
        handler.receive(message);
      });
    handler.on(
      "response", function (response) {
        var responseBuf = swp.buildMessage(response);
        console.log("response: %s", responseBuf.toString());
        stream.write(responseBuf);
      });
    stream.on(
      "data", function (data) {
        parser.execute(data);
      });
    stream.on(
      "end", function () {
        // FIXME: notify handler -> executive
        // TBD: destroy the handler
        handler.removeAllListeners("response");
      });
  });
swankServer.listen(process.argv[2] || 4005, process.argv[3] || "localhost");

function BrowserRemote (clientInfo, client) {
  var userAgent = ua.recognize(clientInfo.userAgent);
  this.name = userAgent.replace(/ /g, "") + (clientInfo.address ? (":" + clientInfo.address) : "");
  this._prompt = userAgent.toUpperCase().replace(/ /g, '-');
  this.pendingRequests = {};
  this.client = client;
  this.client.on(
    "message", function(m) {
      // TBD: handle parse errors
      // TBD: validate incoming message (id, etc.)
      m = JSON.parse(m);
      if (m.op !== "ping") // don't show pings
        console.log("message from browser: %s", JSON.stringify(m));
      switch(m.op) {
      case "output":
        this.output(m.str);
        break;
      case "result":
        if (!this.pendingRequests.hasOwnProperty(m.id)) {
          console.log("WARNING: late result response from the browser");
          break;
        }
        delete this.pendingRequests[m.id];
        if (m.error) {
          this.output(m.error + "\n");
          this.sendResult(m.id, []);
        } else
          this.sendResult(m.id, m.values);
        this.sweepRequests();
        break;
      case "ping":
        this.client.send(JSON.stringify({ "pong": m.id }));
        break;
      default:
        console.log("WARNING: cannot interpret the client message");
      }
    }.bind(this));
    this.client.on(
      "disconnect", function() {
        console.log("client disconnected: %s", this.id());
        this.disconnect();
      }.bind(this));
}

util.inherits(BrowserRemote, swh.Remote);

BrowserRemote.prototype.REQUEST_TIMEOUT = 3000;

BrowserRemote.prototype.sweepRequests = function sweepRequests (all) {
  Object.keys(this.pendingRequests).forEach(
    function (id) {
      if (all || this.pendingRequests[id] < new Date().getTime() - this.REQUEST_TIMEOUT) {
        console.log("request %s didn't finish", id);
        this.sendResult(id, []);
        delete this.pendingRequests[id];
      }
    }, this);
};

BrowserRemote.prototype.prompt = function prompt () {
  return this._prompt;
};

BrowserRemote.prototype.kind = function kind () {
  return "browser";
};

BrowserRemote.prototype.id = function id () {
  return this.name;
};

BrowserRemote.prototype.evaluate = function evaluate (id, str) {
  this.client.send(JSON.stringify({ "id": id, "code": str }));
  this.pendingRequests[id] = new Date().getTime();
};

BrowserRemote.prototype.completion = function completion (id, str) {
  this.client.send(JSON.stringify({ "id": id, "completion": str }));
  this.pendingRequests[id] = new Date().getTime();
};

BrowserRemote.prototype.disconnect = function disconnect () {
  this.sweepRequests(true);
  swh.Remote.prototype.disconnect.call(this);
};

// proxy code from http://www.catonmat.net/http-proxy-in-nodejs

function HttpListener (cfg) {
  this.config = cfg;
}

HttpListener.prototype.clientVersion = "0.1";

HttpListener.prototype.cachedFiles = {};

HttpListener.prototype.clientFiles = {
  'json2.js': 'client/json2.js',
  'stacktrace.js': 'client/stacktrace.js',
  'swank-js.js': 'client/swank-js.js',
  'load.js': 'client/load.js',
  'swank-js-inject.js': 'client/swank-js-inject.js',
  'browser-tests.js': 'client/browser-tests.js',
  'test.html': 'client/test.html',
  'completion.js': 'completion.js'
};

HttpListener.prototype.types = {
  html: "text/html; charset=utf-8",
  js: "text/javascript; charset=utf-8"
};

HttpListener.prototype.scriptBlock =
    new Buffer('<script type="text/javascript" src="/swank-js/swank-js-inject.js"></script>');

HttpListener.prototype.findClosingTag = function findClosingTag (buffer, name) {
  // note: this function is suitable for <head> and <body> tags,
  // because they don't contain any repeating letters, but
  // it will not work for tags that have such letters
  var chars = [];
  var endChar = ">".charCodeAt(0);
  name = "</" + name.toLowerCase();
  for (var i = 0; i < name.length; ++i)
    chars.push(name.charCodeAt(i));
  var A_CODE = "A".charCodeAt(0), Z_CODE = "Z".charCodeAt(0), CODE_INC = "a".charCodeAt(0) - A_CODE;
  function codeToLower (x) {
    return x >= A_CODE && x <= Z_CODE ? x + CODE_INC : x;
  }
  for (i = 0; i < buffer.length - chars.length - 1;) {
    var found = true;
    if (buffer[i++] != chars[0]) // note: no lowercasing for matching against '<'
      continue;

    for (var j = 1; j < chars.length; ++j, ++i) {
      if (codeToLower(buffer[i]) != chars[j]) {
        found = false;
        break;
      }
    }
    if (found) {
      for (var k = i; k < buffer.length; ++k) {
        if (buffer[k] == endChar)// note: no lowercasing for matching against '>'
          return i - chars.length;
      }
    }
  }
  return -1;
};

HttpListener.prototype.injectScripts = function injectScripts (buffer, url) {
  var p = this.findClosingTag(buffer, "head");
  if (p < 0) {
    p = this.findClosingTag(buffer, "body");
    if (p < 0) {
      // html blocks without head / body tags aren't that uncommon
      // console.log("WARNING: unable to inject script block: %s", url);
      return buffer;
    }
  }
  var newBuf = new Buffer(buffer.length + this.scriptBlock.length);
  buffer.copy(newBuf, 0, 0, p);
  this.scriptBlock.copy(newBuf, p, 0);
  buffer.copy(newBuf, p + this.scriptBlock.length, p);
  return newBuf;
};

HttpListener.prototype.proxyRequest = function proxyRequest (request, response) {
  var self = this;
  this.config.get(
    "targetUrl",
    function (targetUrl) {
      self.doProxyRequest(targetUrl, request, response);
    });
};

HttpListener.prototype.doProxyRequest = function doProxyRequest (targetUrl, request, response) {
  var self = this;
  var headersSent = false;
  var done = false;

  var hostname = DEFAULT_TARGET_HOST;
  var port = DEFAULT_TARGET_PORT;
  var parsedUrl = null;
  try {
    parsedUrl = url.parse(targetUrl);
  } catch (e) {}
  if (parsedUrl && parsedUrl.hostname) {
    hostname = parsedUrl.hostname;
    port = parsedUrl.port ? parsedUrl.port - 0 : 80;
  }

  request.headers["host"] = hostname + (port == 80 ? "" : ":" + port);
  delete request.headers["accept-encoding"]; // we don't want gzipped pages, do we?

  // note on http client error handling:
  // http://rentzsch.tumblr.com/post/664884799/node-js-handling-refused-http-client-connections
  var proxy = http.createClient(port, hostname);
  proxy.addListener(
    'error', function handleError (e) {
      console.log("proxy error: %s", e);
      if (done)
        return;
      if (headersSent) {
        response.end();
        return;
      }
      response.writeHead(502, {'Content-Type': 'text/plain; charset=utf-8'});
      response.end("swank-js: unable to forward the request");
  });

  console.log("PROXY: %s %s", request.method, request.url);
  var proxyRequest = proxy.request(request.method, request.url, request.headers);

  proxyRequest.addListener(
    'response', function (proxyResponse) {
      var contentType = proxyResponse.headers["content-type"];
      var statusCode = proxyResponse.statusCode;
      console.log("==> status %s", statusCode);
      var headers = {};
      for (k in proxyResponse.headers) {
        if (proxyResponse.headers.hasOwnProperty(k))
          headers[k] = proxyResponse.headers[k];
      }
      var chunks = proxyResponse.statusCode == 200 && contentType && /^text\/html\b|^application\/xhtml\+xml/.test(contentType) ?
        [] : null;
      if (chunks === null) {
        // FIXME: without this, there were problems with redirects.
        // I don't quite understand why...
        response.writeHead(statusCode, headers);
        headersSent = true;
      }
      proxyResponse.addListener(
        'data', function (chunk) {
          if (chunks !== null) {
            chunks.push(chunk);
            return;
          }
          if (!headersSent) {
            response.writeHead(statusCode, headers);
            headersSent = true;
          }
          response.write(chunk, 'binary');
        });
      proxyResponse.addListener(
        'end', function() {
          if (chunks !== null) {
            console.log("^^MOD: %s %s", request.method, request.url);
            var buf = new Buffer(chunks.reduce(function (s, chunk) { return s += chunk.length; }, 0));
            var p = 0;
            chunks.forEach(
              function (chunk) {
                chunk.copy(buf, p, 0);
                p += chunk.length;
              });
            buf = self.injectScripts(buf, request.url);
            headers["content-length"] = buf.length;
            response.writeHead(statusCode, headers);
            headersSent = true;
            response.write(buf, 'binary');
          } else if (!headersSent) {
            response.writeHead(statusCode, headers);
            headersSent = true;
          }

          response.end();
          done = true;
        });
    });
  request.addListener(
    'data', function(chunk) {
      proxyRequest.write(chunk, 'binary');
    });
  request.addListener(
    'end', function() {
      proxyRequest.end();
    });
};

HttpListener.prototype.getBookMarklets = function getBookMarklets() {
	//TBD: Allow IPv6 Bookmarklets?
	var ifaces = require('os').networkInterfaces();
	var ips = [];

	//TBD: This sucks and can probably be done much better in a functional style
	for (dev in ifaces) {
		console.log ("found device "+dev);
		ifaces[dev].forEach(function(details) {
			console.log(details);
			if (details.family=='IPv4') {
				ips.push(details.address);
			}
		});
	}

	//console.log(ips);

	//TODO: the port is a magic number.  Needs to stop being magic.
	var out = ips.map(function(ip){
	    var bookmarklet = escape("(function(d){window.swank_server='http://"+ip+":8009/';if(!d.getElementById('swank-js-inj')){var h=d.getElementsByTagName('head')[0],s=d.createElement('script');s.id='swank-js-inj';s.type='text/javascript';s.src=swank_server+'swank-js/swank-js-inject.js';h.appendChild(s);}})(document);");
		return '<li><a href="javascript:'+bookmarklet +'"> Connet to slime on '+ip+'</a><br/>javascript:'+bookmarklet+'</li>';			   
	});
	//console.log(out.join('\n'));
	return out.join('\n');
}

HttpListener.prototype.sendCachedFile = function sendCachedFile (req, res, path) {
  if (req.headers['if-none-match'] == this.clientVersion) {
    res.writeHead(304);
    res.end();
  } else {
    // sorry for this, but there is only one replacement, and only one file to replace, so
    // for now... some bad code.  But if there needs to be a new replacement, or replacements
    // on more then one file, this'll need updating.
	var out = ((path == 'client/test.html') ? (this.cachedFiles[path].content+'').replace('<!--[[bookmarklets]]-->',this.getBookMarklets())
			                                : this.cachedFiles[path].content);
	//console.log(out);
	//TBD: Remove the setting of the length header earlier on in the process. 
    this.cachedFiles[path].headers['Content-Length'] = Buffer(out).length;
    res.writeHead(200, this.cachedFiles[path].headers);
    res.end(out, this.cachedFiles[path].encoding);
  }
};

HttpListener.prototype.notFound = function notFound (res) {
  res.writeHead(404, {'Content-Type': 'text/plain; charset=utf-8'});
  res.end("file not found");
};

HttpListener.prototype.serveClient = function serveClient(req, res) {
  var self = this;
  var path = url.parse(req.url).pathname, parts, cn;
  // console.log("%s %s", req.method, req.url);
  if (path && path.indexOf("/swank-js/") != 0) {
    console.log("--> proxy");
    this.proxyRequest(req, res);
    return;
  }


  console.log("--> internal");
  var file = path.substr(1).split('/').slice(1);
  var localPath = this.clientFiles[file];
  if (req.method == 'GET' && localPath !== undefined) {
    // TBD: reenable caching, check datetime of the file
    // if (path in this.cachedFiles){
    //   this.sendCachedFile(req, res, path);
    //   return;
    // }

    fs.readFile(
      __dirname + '/' + localPath, function(err, data) {
        if (err) {
          console.log("error: %s", err);
          self.notFound(res);
        } else {
          var ext = localPath.split('.').pop();
          self.cachedFiles[localPath] = {
			// right now there is no difference between cached files
            // and files inside of the client dir.  That should probably change 
            // soon.
            headers: {
              'Content-Length': data.length,
              'Content-Type': self.types[ext],
              'ETag': self.clientVersion
            },
            content: data,
            encoding: ext == 'swf' ? 'binary' : 'utf8'
          };
          self.sendCachedFile(req, res, localPath);
        }
      });
  } else {
    console.log("bad request for /swank-js/ path");
    this.notFound(res);
  }
};

var httpListener = new HttpListener(cfg);
var httpServer = http.createServer(httpListener.serveClient.bind(httpListener));

httpServer.listen(8009);
io = io.listen(httpServer);

io.sockets.on(
  "connection", function (client) {
    // new client is here!
    console.log("client connected");
    function handleHandshake (message) {
      message = JSON.parse(message);
      client.removeListener("message", handleHandshake);
      if (!message.hasOwnProperty("op") || message.op != "handshake")
        console.warn("WARNING: skipping pre-handshake message: %j", message);
      else {
        var address = null;
        if (client.connection && client.connection.remoteAddress)
          address = client.connection.remoteAddress || "noaddress";
        var remote = new BrowserRemote({ address: address, userAgent: message.userAgent || "" }, client);
        executive.attachRemote(remote);
        console.log("added remote: %s", remote.fullName());
      }
    };
    client.on("message", handleHandshake);
  });

// TBD: handle reader errors

// function location determination:
// for code loaded from scripts: direct (if possible)
// for 'compiled' code: load the code by adding <script> tag loaded from the swank-js' webserver, its name should encode the real path and line offset
// for code entered via REPL: none
// PREPROCESS STACK TRACES!!!
// https://github.com/emwendelin/Javascript-Stacktrace
// ALSO: http://blog.yoursway.com/2009/07/3-painful-ways-to-obtain-stack-trace-in.html -- onerror in ie gives the innermost frame
// it should be also possible to 'soft-trace' functions so that they extend Exception objects with caller info as it passes through them
// TBD: unix domain sockets, normal slime startup
// TBD: http request logging (for specific remote)
// TBD: sudden disconnections (flashsocket), sometimes after lots of output (?) --
// Error: You are trying to call recursively into the Flash Player which is not allowed. In most cases the JavaScript setTimeout function, can be used as a workaround.
// TBD: autoreconnect + connection error handling
// ALSO: are htmlfile, jsonp-polling modes etc supposed to disconnect after each message?
// TBD: add SwankJS scripts to all passing html pages (into <head> or <body>)
// TBD: it should be possible to serve local files instead of proxying
// (maybe using https://github.com/felixge/node-paperboy )
// TBD: handle edge case: new sticky remote connects, old sticky remote disconnects
// (late disconnect) - as of now, swank-js switches to node.js, but it should
// instead upon remote detachment see whether another remote with the same name
// is available
// TBD: handle/add X-Forwarded-For headers
// TBD: fix all assert calls: we need (actual, expected) not (expected, actual)
// TBD: invoke SwankJS.setup() only when DOM is ready (at least in IE)
// TBD: timeouts for browser requests
